src/esp/gfx/TextRenderer.cpp

```diff
@@ -0,0 +1,141 @@
+
+// This is to import the "resources" at runtime. When the resource is
+// compiled into static library, it must be explicitly initialized via this
+// macro, and should be called *outside* of any namespace.
+static void importFontResources() {
+  CORRADE_RESOURCE_INITIALIZE(FontResources)
+}
+
+namespace esp {
+namespace gfx {
+
+const std::string defaultCharacters =
+    "abcdefghijklmnopqrstuvwxyz"
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    "0123456789?!:;,.# ";
+
+constexpr float TextRenderer::locationLookup_[NUM_LOCATION][2];
+
+TextRenderer::TextRenderer(float aspectRatio, float fontSize, int cacheSize)
+    : TextRenderer(aspectRatio, fontSize, cacheSize, defaultCharacters) {}
+
+TextRenderer::TextRenderer(float aspectRatio,
+                           float fontSize,
+                           int cacheSize,
+                           const std::string& characters)
+    : cache_(Mn::Vector2i{cacheSize}) {
+  if (!Cr::Utility::Resource::hasGroup("fonts")) {
+    importFontResources();
+  }
+  CORRADE_ASSERT(
+      fontSize > 0,
+      "TextRenderer::TextRenderer: font size" << fontSize << "is illegal.", );
+  CORRADE_ASSERT(
+      cacheSize > 0,
+      "TextRenderer::TextRenderer: cache size" << cacheSize << "is illegal.", );
+
+  font_ = fontManager_.loadAndInstantiate("StbTrueTypeFont");
+  CORRADE_ASSERT(
+      font_,
+      "TextRenderer::TextRenderer: cannot load and instantiate the font", );
+
+  Cr::Utility::Resource rs("fonts");
+  std::string fontName = "SourceSansPro-SemiBold.ttf";
+  if (!font_->openData(rs.getRaw(fontName), fontSize)) {
+    Cr::Utility::Fatal{-1} << "TextRenderer::TextRenderer: The font" << fontName
+                           << "does not exist.";
+  }
+
+  font_->fillGlyphCache(cache_, characters);
+
+  updateAspectRatio(aspectRatio);
+}
+
+size_t TextRenderer::createRenderer(float onScreenCharacterSize,
+                                    size_t glyphCount,
+                                    TextLocation location) {
+  CORRADE_ASSERT(onScreenCharacterSize > 0,
+                 "TextRenderer::createRenderer: the character size"
+                     << onScreenCharacterSize << "is illegal.",
+                 0);
+
+  CORRADE_ASSERT(glyphCount > 0,
+                 "TextRenderer::createRenderer: the capacity" << glyphCount
+                                                              << "is illegal.",
+                 0);
+
+  Mn::Text::Alignment alignment = Mn::Text::Alignment::TopLeft;
+  ;
+  switch (location) {
+    case TOP_LEFT:
+      // has been initialized above
+      break;
+    case TOP_RIGHT:
+      alignment = Mn::Text::Alignment::TopRight;
+      break;
+    default:
+      Cr::Utility::Fatal{-1} << "TextRenderer::createRenderer: The location"
+                             << "is not supported.";
+      break;
+  }
+  text_.emplace_back(
+      Cr::Containers::Pointer<Mn::Text::Renderer2D>(new Mn::Text::Renderer2D(
+          *font_, cache_, onScreenCharacterSize, alignment)));
```

Comment:
Thank you for your explanation in the previous comments.
Now I understand e.g., the TopLeft means the topleft corner of the text is located at the origin. And more importantly, the origin (0, 0) is originally at the center of the window. So the TopLeft is (-1, 1), TopRight is (1, 1), BottomLeft is (-1, -1) and BottomRight is (1, -1).